[postgresql_ext] Add support for requiring a specific extension version

strk opened the pull request:
##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

postgresql_ext module

##### SUMMARY
Briefly tested manually

strk committed:
[postgresql_ext] Add support for requiring a specific extension version

strk committed:
Fix indent and add version_added for new option

strk commented:
##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

postgresql_ext module

strk committed:
Fix syntax (now tested)

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_ext.py:
```diff
@@ -98,9 +104,12 @@ class NotSupportedError(Exception):
-    cursor.execute(query, {'ext': ext})
+def ext_exists(cursor, ext, ver=None):
+    if ver:
+        query = "SELECT * FROM pg_extension WHERE extname=%(ext)s and extversion=$(ver)s"
```
It seems a `ProgrammingError: syntax error at or near "$"` occurs.

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_ext.py:
```diff
@@ -111,12 +120,20 @@ def ext_delete(cursor, ext):
     if not ext_exists(cursor, ext):
-        query = 'CREATE EXTENSION "%s"' % ext
+        if ver:
+            query = 'CREATE EXTENSION "%s" VERSION "%s"' % ( ext, ver)
```
Don't use string parameters interpolation (`%`) to pass variables to a SQL query string ([see](http://initd.org/psycopg/docs/usage.html#the-problem-with-the-query-parameters)). Same comment applies to queries below.

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_ext.py:
```diff
@@ -132,6 +149,7 @@ def main():
             port=dict(default="5432"),
             db=dict(required=True),
             ext=dict(required=True, aliases=['name']),
+            version=dict(default="", aliases=['version']),
```
`DOCUMENTATION` metadata states that default value for `version` is `None`, not an empty string. Documentation and code must match.

pilou- replied:
`aliases` is useless (the same name is used).

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_ext.py:
```diff
@@ -175,15 +194,15 @@ def main():
+                changed = ext_exists(cursor, ext, ver)
         else:
             if state == "absent":
                 changed = ext_delete(cursor, ext)
```
`version` must be checked for deletion operation. Moreover when `check_mode` is enabled, `version` is checked: when `check_mode` is enabled one particular version is tested, when `check_mode` is disabled any version is deleted.

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_ext.py:
```diff
@@ -34,6 +34,12 @@
     default: null
+  version:
+    description:
+      - version of the extension to add or remove (or update to)
```
Add something like `when not specified, any version will match`.

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_ext.py:
```diff
@@ -142,6 +160,7 @@ def main():
 
     db = module.params["db"]
     ext = module.params["ext"]
+    ver = module.params["version"]
```
I would prefer a variable named `version` to `ver`.

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_ext.py:
```diff
@@ -111,12 +120,20 @@ def ext_delete(cursor, ext):
+        if ver:
+            if not ext_exists(cursor, ext, ver):
+                query = 'ALTER EXTENSION "%s" UPDATE TO "%s"' % ( ext, ver )
+            cursor.execute(query)
```
`query` is unset when `ext_exists` returns `True`.

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_ext.py:
```diff
@@ -34,6 +34,12 @@
+  version:
+    description:
+      - version of the extension to add or remove (or update to)
+    required: false
```
`false` and `null` are default values for `required` and `default` respectively, both fields could be removed.

pilou- reviewed (commented):
Please, add some tests (unit or integration). See http://docs.ansible.com/ansible/dev_guide/testing.html#types-of-tests. There are some PostgreSQL related integration tests in 'test/integration/targets/postgresql/tasks'.

strk committed:
Fix ProgrammingError and avoid use string param interpolation for sql

strk committed:
Remove useless alias

strk committed:
Improve documentation about version

strk committed:
Fix use of undefined query when creating already-existing version

strk commented:
I've addressed almost all your comments, except:

 - unit test (not done yet)
 - removal of "required" and "default" values (they are used elsewhere
   too, in the same module, and they should not harm)

As for the unchecked version in "absent" case that's by design, I now
documented it.

pilou- commented:
`changed` value is inconsistent, an update is required.

If `version` parameter is specified and `state` parameter is `absent`:
* when `check_mode` is enabled,  `version` is ignored: module returns `changed=True` if any version exists;
* when `check_mode` is disabled, `version` isn't ignored: module returns `changed=True` only if the specified version exists, otherwise module returns `changed=False`.

You might want to test value of `check_mode` inside `ext_create`/`ext_delete` methods (it would require add a new parameter to these methods) rather than testing `check_mode` value in the caller.

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_ext.py:
```diff
@@ -111,12 +123,20 @@ def ext_delete(cursor, ext):
+            if not ext_exists(cursor, ext, ver):
+                query = 'ALTER EXTENSION "%(ext)s" UPDATE TO "%(ver)s"'
+                cursor.execute(query, {'ext': ext, 'ver': ver})
+            return True
```
Indentation level of `return` statement need to be increased.

strk committed:
Do not return changed=True when requested ext version already exist

strk commented:
Does 939a5ad75 address all raised issues @Pilou ?

pilou- commented:
@strk what about the [comment](https://github.com/ansible/ansible/pull/26737#issuecomment-315613520) related to `check_mode` ?

strk commented:
The `check_mode` behavior seems expected to me. `absent` means no version must exist.
A single version may exist in a single database, so `absent` means presence of any version requires a change (and `version` is just not used in that case). This was explained in https://github.com/ansible/ansible/pull/26737#issuecomment-315518386 and documented in https://github.com/ansible/ansible/pull/26737/commits/5dd8b28cb35296db7227acdc9d5084f16a4010cb

pilou- commented:
The module documentation states:
> version of the extension to add or remove (or update to), **has no effect** when requested state is "absent"

but when `check_mode` is enabled and `state` is `absent`, `version` parameter has an effect: value of this parameter is read and [used](https://github.com/strk/ansible/blob/939a5ad7544cca45463a0e3a6f4f691f738560b1/lib/ansible/modules/database/postgresql/postgresql_ext.py#L202).

When `state` is `absent`, it seems more logical to remove any version when `version` is not specified and to remove the chosen version only when `version` is specified. It would allow the user to remove a specific version only.

strk committed:
Fix would-be-changed in check mode with state=absent

strk commented:
Good catch !
Does 737cdcae6adea5e747ef1995030bc216d999e536 address it ?

pilou- commented:
@strk yes

pilou- commented:
@strk just for you to know, Ansible 2.4 feature freeze for community module is planned for 29 August 2017.

strk committed:
Attempt to provide an integration test

WARNING: does not really test version, and assumes "postgis"
extension version 2.2.0 and 2.3.0 is available

strk commented:
I've rebased to develop branch to fix conflicts and pushed a blind attempt at an integration test.
Hadn't found any integration test for the postgresql_ext module itself (was there any?) so added one.
The test assumes "postgis" extension is available and is available in both versions 2.2.0 and 2.3.0 - it is highly unlikely that this is the case so instructions about having the build both installing them are welcome.

pilou- committed:
PostgreSQL tests: setup a test extension

pilou- committed:
'state' must be 'absent' in order to remove extension

pilou- committed:
Fix assert value

pilou- committed:
Test deletion even when version isn't set

pilou- committed:
Simplify boolean expression used in assert tasks

pilou- committed:
PostgreSQL tests: use dummy extension

- dummy extension is easier to install than PostGIS
- 'nil' is not a YAML null value

pilou- committed:
postgresql_ext tests require a database

pilou- committed:
postgresql_ext: identifiers must be quoted

extension name is an identifier, extension version is either an
identifier or a literal string.

- for extension name: use psycopg2.extensions.quote_ident when available,
otherwise check if name is a valid identifier.
- handle extension version using parametrized query.

pilou- commented:
@strk I pushed some commits on top of your commits there: https://github.com/pilou-/ansible/tree/postgresql_ext_version. The updates are:
- instead of using PostGIS, `setup_postgresql` role configure and install a test extension
- some commits are related to identifier quoting/escaping (extension `name` is an identifier which can not be escaped/quoted like parameters).

I was able to execute the integration test suite successfully using: `ansible-test integration --docker centos7 --docker-privileged postgresql` (it's work with `ubuntu1604` and `ubuntu1604py3` docker images too).

Please, could you append commits from my branch to this one ? Something like these commands could be used for that):
1. git checkout postgresql_ext_version
1. git remote add -t postgresql_ext_version pilou https://github.com/pilou-/ansible
1. git fetch pilou
1. git merge --ff-only pilou/postgresql_ext_version
1. git push

mattclay commented:
CI failure in integration tests due to traceback:

```
Traceback (most recent call last):
  File "/tmp/ansible_C7OVXY/ansible_module_postgresql_ext.py", line 179, in main
    db_connection = psycopg2.connect(database=db, **kw)
  File "/usr/lib/python2.7/dist-packages/psycopg2/__init__.py", line 164, in connect
    conn = _connect(dsn, connection_factory=connection_factory, async=async)
OperationalError: FATAL:  database "ansible_db" does not exist
```

strk commented:
Thanks, your branch was merged.
I don't have `ansible-test` on my system, and `apt-cache search`
doesn't find it either. How to obtain it ?

strk committed:
Fix docs not to suggest you can remove specific versions

strk commented:
Another thing, since we're installing the extension ourselves we could
as well test that with no `version` we get the latest installed
version (the one referenced by dummy_extension.control).

Can you add that ?
Current test only checks for verison when given:

      - "not ext_version or ext_version in result.stdout"

pilou- committed:
Test postgresql_ext only if supported by PostgreSQL

SQL extension commands are supported by PostgreSQL 9.1 and later.

pilou- committed:
postgresql_ext: hardcode some extension directories

pilou- committed:
FreeBSD host: pg_ver isn't used

pilou- committed:
Check available version when version isn't a parameter

pilou- committed:
Check that function defined in extension is available

pilou- commented:
Updated (on top of your commits, [here](https://github.com/pilou-/ansible/tree/postgresql_ext_version)):
- available version is checked even when `version` parameter is `None`
- check that function defined by extension is available
- Fix Centos 6 error: don't run test when extension SQL commands aren't available
- Fix FreeBSD failure: hardcode some extension directories

`ansible-test` command is in `test/runner/`. You can use `. hacking/env-setup` in order to add this command to `$PATH`. The easiest way is to use Docker and the command pasted before.
Some documentation is available on [dev_guide](http://docs.ansible.com/ansible/latest/dev_guide/testing_integration.html).

strk commented:
New commits merged, and testsuite successfully run locally.
Thanks !

pilou- commented:
It seems there is a problem with one check on FreeBSD and RHEL 7.4, I need to check again.
```
2017-08-06 17:04:33 TASK [setup_postgresql_db : restart postgresql service] ************************
2017-08-06 17:04:36 changed: [testhost] => {"changed": true, "failed": false, "name": "postgresql", "state": "started"}
2017-08-06 17:04:36 
2017-08-06 17:04:36 TASK [setup_postgresql_db : Check if extension are supported] ******************
2017-08-06 17:51:05 Password: Connection to ec2-34-204-44-196.compute-1.amazonaws.com closed by remote host.
```

pilou- committed:
Fix a typo s/not set/undefined/

pilou- committed:
postgresq_ext: default user isn't always postgres

workaround for undocumented feature: default value of login_user
parameter is 'postgres'.

pilou- committed:
fix usage of default filter

pilou- committed:
Use PostgreSQL admin user instead of root

pilou- committed:
postgresql_ext: fix formatting format

pilou- committed:
postgresql_ext tests: fix check

pilou- commented:
[Branch](https://github.com/pilou-/ansible/tree/postgresql_ext_version) updated, now tests should work on FreeBSD and RHEL 7.4 (both tested locally).

mattclay commented:
CI is failing on FreeBSD and RHEL with the tests hanging at `setup_postgresql_db : Check if extension are supported`.

strk commented:
Fixes for FreeBSD and RHEL 7.4 merged

pilou- commented:
cc @dschep @matburt @nerzhul could you please review this PostgreSQL related pull request? Integration tests have been added and CI is successful :)

dschep reviewed (approved):
shipit :shipit:

pilou- commented:
When `psycopg2.extensions.quote_ident` isn't available (meaning Psycopg2 is older than 2.7), [`postgresql_ext` uses `re` module](https://github.com/ansible/ansible/pull/26737/files#diff-7771d17a1549cfa9ce8727aab21a0072R101) in order to allow unquoted identifiers only.

Because Psycopg2 is [quite recent](https://github.com/psycopg/psycopg2/releases/tag/2_7) and `PQescapeIdentifier` is available since [libpq 9.0](https://www.postgresql.org/docs/9.0/static/libpq-exec.html#LIBPQ-PQESCAPEIDENTIFIER), I propose (#30828):
- to use `PQescapeIdentifier` when available
- when both `psycopg2.extensions.quote_ident` and `PQescapeIdentifier` aren't available, then unquoted identifiers only are allowed.
- remove buggy (#22568) method `pg_quote_identifier` from [module_utils.database.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/database.py#L113)

@strk , @matburt , @nerzhul, @dschep : what do you think ?

strk commented:
Sorry I've no idea what you're talking about, I don't remember ever adding code to deal with codes in my original PR. That said, quoted identifiers shoudn't be too hard to implement, just double any `"` and wrap the result in `"`.

pilou- commented:
@stk in [prepare_ext_query method](https://github.com/ansible/ansible/pull/26737/files#diff-7771d17a1549cfa9ce8727aab21a0072R122) `psycopg2.extensions.quote_ident` is used when Psycopg2 >= 2.7.

strk commented:
@pilou- I see you added that code: https://github.com/ansible/ansible/pull/26737/commits/8164b35d8d73e451b17a0cf3f1db7f3ee3ca3c2d

I agree with generalizing the identifiers quoting function, and again it should be fairly simple to do.
But #30828 is too big to review... 

Is the test in this PR checking for MixedCase identifiers of extension names ?

pilou- commented:
Branch of #30828 pull request is based on this pull request branch, only three commits have been added:
- b36bfb6db4691b5711a39a418aef0aa785d43bd0: code allowing to escape identifiers
- a392aceda42646c6b0bcc2551c1e91c27fa165b0: unit tests
- f3230a8122c95a057e6bf2253328f2178c651bdc: using `escape_identifier_cursor` in `postgresql_ext`

The purpose of #30828 is to show `escape_identifier_cursor` usage and to receive feedback, then I will update #30828 keeping only the first two commits.

Andersson007 commented:
@strk , hey! I'm going to add this option by another PR, would you mind if I add you as a coauthor?

strk commented:
@Andersson007 sure, no problem, thanks for asking

Andersson007 commented:
@strk , thank you, let you know when I make a PR

Andersson007 commented:
I made one [#58381](https://github.com/ansible/ansible/pull/58381) instead
